src/container/srv_target.c

```diff
@@ -2548,39 +2667,54 @@ ds_cont_tgt_ec_eph_query_ult(void *data)
 					}
 				}
 
-				if (!is_failed_tgts)
-					min_eph = min(min_eph, ec_eph->ce_ephs[i]);
+				if (!is_failed_tgts) {
+					min_ec_agg_eph = min(min_ec_agg_eph,
+							     ec_eph->cte_ec_agg_ephs[i]);
+					min_stable_eph = min(min_stable_eph,
+							     ec_eph->cte_stable_ephs[i]);
+				}
 			}
 
-			if (min_eph == 0 || min_eph == DAOS_EPOCH_MAX ||
-			    min_eph <= ec_eph->ce_last_eph) {
-				if (min_eph > 0 && min_eph < ec_eph->ce_last_eph)
-					D_ERROR("ignore for now "DF_X64" < "DF_X64
-						" "DF_UUID"\n", min_eph, ec_eph->ce_last_eph,
-						DP_UUID(ec_eph->ce_cont_uuid));
+			if (min_ec_agg_eph == 0 || min_ec_agg_eph == DAOS_EPOCH_MAX ||
+			    min_stable_eph == 0 || min_stable_eph == DAOS_EPOCH_MAX ||
+			    (min_ec_agg_eph <= ec_eph->cte_last_ec_agg_epoch &&
+			     min_stable_eph <= ec_eph->cte_last_stable_epoch)) {
+				if (min_ec_agg_eph > 0 && min_stable_eph > 0 &&
+				    min_ec_agg_eph <= ec_eph->cte_last_ec_agg_epoch &&
+				    min_stable_eph <= ec_eph->cte_last_stable_epoch)
+					D_ERROR("ignore for now "DF_X64" <= "DF_X64
+						", "DF_X64" <= "DF_X64 ", "DF_UUID"\n",
+						min_ec_agg_eph, ec_eph->cte_last_ec_agg_epoch,
+						min_stable_eph, ec_eph->cte_last_stable_epoch,
+						DP_UUID(ec_eph->cte_cont_uuid));
 				else
 					D_DEBUG(DB_MD, "Skip eph "DF_X64"/"DF_X64
-						" "DF_UUID"\n", min_eph, ec_eph->ce_last_eph,
-						DP_UUID(ec_eph->ce_cont_uuid));
+						", "DF_X64"/"DF_X64", "DF_UUID"\n",
+						min_ec_agg_eph, ec_eph->cte_last_ec_agg_epoch,
+						min_stable_eph, ec_eph->cte_last_stable_epoch,
+						DP_UUID(ec_eph->cte_cont_uuid));
 				continue;
 			}
 
-			D_DEBUG(DB_MD, "Update eph "DF_X64" "DF_UUID"\n",
-				min_eph, DP_UUID(ec_eph->ce_cont_uuid));
-			rc = cont_iv_ec_agg_eph_update(pool->sp_iv_ns, ec_eph->ce_cont_uuid,
-						       min_eph);
-			if (rc == 0)
-				ec_eph->ce_last_eph = min_eph;
-			else
+			D_DEBUG(DB_MD, "Update ec_agg_eph "DF_X64", stable_eph "DF_X64", "
+				DF_UUID"\n", min_ec_agg_eph, min_stable_eph,
+				DP_UUID(ec_eph->cte_cont_uuid));
+			rc = cont_iv_track_eph_update(pool->sp_iv_ns, ec_eph->cte_cont_uuid,
+						      min_ec_agg_eph, min_stable_eph);
+			if (rc == 0) {
+				ec_eph->cte_last_ec_agg_epoch = min_ec_agg_eph;
+				ec_eph->cte_last_stable_epoch = min_stable_eph;
+			} else {
 				D_INFO(DF_CONT": Update min epoch: %d\n",
-				       DP_CONT(pool->sp_uuid, ec_eph->ce_cont_uuid), rc);
+				       DP_CONT(pool->sp_uuid, ec_eph->cte_cont_uuid), rc);
+			}
 		}
 		D_FREE(failed_tgts);
 yield:
 		if (dss_ult_exiting(pool->sp_ec_ephs_req))
 			break;
 
-		sched_req_sleep(pool->sp_ec_ephs_req, EC_TGT_AGG_INTV);
+		sched_req_sleep(pool->sp_ec_ephs_req, EC_TGT_EPH_QUERY_INTV);
```

Comment:
That means per-5 seconds, current engine will report its local stable epoch (per-ppol) to the leader, right?

Reply:
yes, reports each container's local stable epoch and ec agg epoch to leader per container.